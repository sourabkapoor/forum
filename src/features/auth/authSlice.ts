import { createSlice } from "@reduxjs/toolkit";
import { RootState } from "../../redux/store";

interface AuthState {
  isLoggedIn: boolean;
  signUpModalState: boolean;
  signInModalState: boolean;
}

const initialState: AuthState = {
  isLoggedIn: false,
  signUpModalState: false,
  signInModalState: false,
};

export const selectAuthStatus = (state: RootState) => state.auth.isLoggedIn;

export const selectSignUpModalState = (state: RootState) =>
  state.auth.signUpModalState;

export const selectSignInModalState = (state: RootState) =>
  state.auth.signInModalState;

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    setLoginStatus: (state, action) => {
      state.isLoggedIn = action.payload;
    },
    toggleSignUpModal: (state) => {
      state.signUpModalState = !state.signUpModalState;
    },
    toggleSignInModal: (state) => {
      state.signInModalState = !state.signInModalState;
    },
  },
});

export const { setLoginStatus, toggleSignUpModal, toggleSignInModal } =
  authSlice.actions;

export default authSlice.reducer;