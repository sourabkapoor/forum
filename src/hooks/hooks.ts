import {
  collection,
  query,
  orderBy,
  getDocs,
  DocumentData,
  where,
  onSnapshot,
} from "firebase/firestore";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import { db } from "../firebase";
import { AppDispatch } from "../redux/store";
import { RootState } from "../redux/store";

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useFilter = () => {
  async function filterNew() {
    const postsDocsRef = collection(db, "posts");

    const q = query(postsDocsRef, orderBy("createdAt", "desc"));

    const data = await getDocs(q);
    return data.docs;
  }

  async function filterTop() {
    const postsDocsRef = collection(db, "posts");

    const q = query(postsDocsRef, orderBy("voteStatus", "desc"));

    const data = await getDocs(q);
    return data.docs;
  }

  async function filterRising() {
    const postsDocsRef = collection(db, "posts");

    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);

    const q = query(
      postsDocsRef,
      where("createdAt", ">=", yesterday)
    );

    const data = await getDocs(q);

    const newData = data.docs.sort((a, b) => a.data().voteStatus - b.data().voteStatus)
    return newData;
  }

  return {
    filterNew,
    filterTop,
    filterRising,
  };

};

export const useSubredditFilter = (subredditName: string | undefined) => {
  const filterNew = async () => {
    const postsDocsRef = collection(db, "posts");

    const q = query(
      postsDocsRef,
      where("subredditName", "==", subredditName),
      orderBy("createdAt", "desc")
    );
    const data = await getDocs(q);
    return data.docs;
  };

  const filterTop = async () => {
    const postsDocsRef = collection(db, "posts");

    const q = query(
      postsDocsRef,
      where("subredditName", "==", subredditName),
      orderBy("voteStatus", "desc")
    );

    const data = await getDocs(q);
    return data.docs;
  };

  const filterRising = async () => {
    const postsDocsRef = collection(db, "posts");

    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);

    const q = query(
      postsDocsRef,
      where("subredditName", "==", subredditName),
      where("createdAt", ">=", yesterday)
    );

    const data = await getDocs(q);

    const newData = data.docs.sort(
      (a, b) => a.data().voteStatus - b.data().voteStatus
    );
    return newData;
  };

  return {
    filterNew,
    filterTop,
    filterRising,
  };
};

