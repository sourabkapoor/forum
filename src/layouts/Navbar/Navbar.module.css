.header {
  /* background-color: rgb(36, 37, 38); */
  /* background-color: rgb(36, 37, 38); */
  height: 48px;
  /* border-bottom: 1px solid rgb(57, 58, 59); */
  box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
}

.header__content {
  max-width: 97%;
  display: flex;
  align-items: center;
  padding-left: 20px;
  padding-right: 20px;
  height: 100%;
}

.header__left {
  display: flex;
  align-items: center;
  flex-grow: 0;
}

.header__dropdown-profile {
  width: 100%;
}

.header__left:hover {
  cursor: pointer;
}

.header__links {
  display: flex;
  align-items: center;
  font-weight: 900;
  color: blueviolet;
}

.header__dropdown {
  display: flex;
  align-items: center;
  gap: 5px;
  max-width: 270px;
  justify-content: space-between;
  border: 1px solid #cecece;
  padding: 0.5rem 0.25rem;
  box-sizing: border-box;
  color: black;
  border-radius: 4px;
  margin: 0 12px;
}

.header__dropdown:hover {
  /* border: 1px solid #343536; */
  box-sizing: border-box;
  
}

.header__dropdown-left {
  display: flex;
  align-items: center;
  color: black;
  gap: 8px;
}

.header__dropdown-icon {
  width: 20px;
  height: 20px;
  color: black;
  line-height: 34px;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.header__dropdown-title {
  display: inline;
  color: black;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
}

.header__dropdown-menu {
  max-width: 270px;
  position: relative;
  overflow: scroll;
}

.header__dropdown-icon {
  display: flex;
  align-items: center;
}

.header__logo {
  width: 32px;
  height: 32px;
  padding-right: 8px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: blueviolet;
}

.header__logo-name {
  width: 57px;
  height: 18px;
  margin-right: 20px;
}

.header__container {
  background-color: #e2e2e2;
  color: var(--secondary-color);
  margin: 0 auto;
  width: 690px;
  flex-grow: 1;
  border-radius: 10px;
  display: flex;
  height: 36px;
  align-items: center;
  margin-left: 16px;
  margin-right: 16px;
}

.header__search-icon-container {
  padding: 0 9px 0 15px;
  padding-left: 15px;
  padding-right: 9px;
  display: flex;
  align-items: center;
}

.header__search-icon {
  color: gray;
  width: 20px;
  height: 36px;
}

.header__search-input {
  /* background-color: rgb(25, 26, 27); */
  border: none;
  background-color: transparent;
  font-weight: 400;
  color: var(--secondary-color);
  width: 99%;
}

.header__search-input:focus {
  all: unset;
}

.header__right {
  display: inline-flex;
  align-items: center;
  flex-grow: 0;
}

.header__buttons {
  display: flex;
  align-items: center;
}

.header__button {
  padding: 4px 16px;
  min-height: 32px;
  min-width: 32px;
  line-height: 17px;
  text-transform: unset;
  font-weight: 500;
  font-size: 14px;
  border-radius: 9999px;
  border: none;
  box-sizing: border-box;
  text-align: center;
}

.header__button_type_log-in {
  /* background-color: black; */
  color: black;
  border: 1px solid white;
}

.header__button_type_log-in:hover {
  /* background-color: rgb(36, 37, 38); */
  cursor: pointer;
}

.header__button_type_sign-up {
  margin-left: 4px;
}

.header__button_type_sign-up:hover {
  /* background-color: rgb(111, 111, 111); */
  cursor: pointer;
}

.header__search-form {
  width: 100%;
}

@media (max-width: 1110px) {
  .header_buttons {
    width: 100%;
  }
  .header__button {
    width: 85px;
  }
}

@media (max-width: 1070px) {
  .header__logo-name {
    display: none;
  }
}

@media (max-width: 995px) {
  .header__dropdown-title {
    display: none;
  }
}

@media (max-width: 640px) {
  .header__search-input:focus {
    width: 100%;
  }
}
