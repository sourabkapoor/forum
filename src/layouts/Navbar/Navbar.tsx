/* eslint-disable no-restricted-globals */
/* eslint-disable eqeqeq */
import React, { useEffect, useState } from "react";
import styles from "./Navbar.module.css";
import logo from "../../assets/logo.svg";
import { Link, Navigate, useNavigate, useParams } from "react-router-dom";
import logoName from "../../assets/white-logo-name.svg";
import classNames from "classnames";
import { getUser, signIn, signOutUser } from "../../firebase";
import { useAppDispatch, useAppSelector } from "../../hooks/hooks";
import {
  selectAuthStatus,
  toggleSignInModal,
  toggleSignUpModal,
} from "../../features/auth/authSlice";
import ChatBubbleIcon from '@mui/icons-material/ChatBubble';
import { RiSearchLine } from "react-icons/ri";
import { FaRegUser } from "react-icons/fa";
import { IoIosArrowDown } from "react-icons/io";
import Profile from "../Profile/Profile";
import CSSModules from "react-css-modules";
import { AiFillHome } from "react-icons/ai";
import Dropdown from "../Dropdown/Dropdown";
import CommunityDropdown from "../CommunityDropdown/CommunityDropdown";
import ProfileDropdown from "../ProfileDropdown/ProfileDropdown";
import { selectCommunityData } from "../../features/subreddit/subredditSlice";
import NavigationDrawer from "../NavigationDrawer/NavigationDrawer";

type InputEvent = React.ChangeEvent<HTMLInputElement>;

type FormEvent = React.FormEvent<HTMLFormElement>;

interface Props {
  toggleDrawer?: Function;
}

const Navbar: React.FC<Props> = ({toggleDrawer}) => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { name } = useAppSelector(selectCommunityData);
  const isLoggedIn = useAppSelector(selectAuthStatus);

  const params = useParams();

  // const [isSubscriptionsDropdownOpen, setIsSubscriptionsDropdownOpen] =
  //   useState(false);

  const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);

  const [searchInput, setSearchInput] = useState("");

  function handleSearchInput(event: InputEvent) {
    setSearchInput(event.target.value);
  }

  function submitSearch(event: FormEvent) {
    setTimeout(() => {
      navigate(`${searchInput}`);
      setSearchInput("");
      location.reload();
    }, 500);
  }

  function handleProfileDropdown() {
    setIsProfileDropdownOpen((prevValue) => !prevValue);
  }

  // function handleHomeDropdown() {
  //   setIsSubscriptionsDropdownOpen((prevValue) => !prevValue);
  // }

  function handleSignUpClick() {
    dispatch(toggleSignUpModal());
  }

  function handleSignInClick() {
    dispatch(toggleSignInModal());
  }

  return (
    <header styleName="header">
      <div styleName="header__content">
        <div styleName="header__left">
          <Link to="/" styleName="header__links">
            <ChatBubbleIcon styleName="header__logo" />
            Forum
          </Link>
          {/* Dropdown */}
          {isLoggedIn && (
            <div styleName="header__dropdown-container">
              <div styleName="header__dropdown" onClick={() => toggleDrawer && toggleDrawer()}>
                <div styleName="header__dropdown-left">
                  <AiFillHome styleName="header__dropdown-icon" />
                  <h1 styleName="header__dropdown-title">
                    {Object.keys(params).length === 0 ? "Home" : name}
                  </h1>
                </div>
                <div styleName="header__dropdown-right">
                  <IoIosArrowDown styleName="header__dropdown-icon" />
                </div>
              </div>
              {/* <div className="header__dropdown-menu">
                {isSubscriptionsDropdownOpen && (
                    <NavigationDrawer
                    dropdown={isSubscriptionsDropdownOpen}
                    handleHomeDropdown={() => handleHomeDropdown()}/>
                  )}
              </div> */}
            </div>
          )}
        </div>

        {/* Search bar */}
        <div styleName="header__container">
          <div styleName="header__search-icon-container">
            <RiSearchLine styleName="header__search-icon" />
          </div>
          <form onSubmit={submitSearch} styleName="header__search-form">
            <input
              styleName="header__search-input"
              placeholder="Search Forum ie. (r/dogs)"
              onChange={handleSearchInput}
              value={searchInput}
              pattern="(?:^| )(/?r/[a-z]+)"
              title={`${searchInput} is not valid. Please enter a 'r/', followed by a forum name in the format 'r/dogs'`}
            />
          </form>
        </div>
        
        <div styleName="header__right">
          <div styleName="header__buttons">
            {isLoggedIn === false && (
              <>
                <button
                  onClick={handleSignInClick}
                  styleName="header__button header__button_type_log-in"
                >
                  Log In
                </button>
                <button
                  onClick={handleSignUpClick}
                  styleName="header__button header__button_type_sign-up"
                >
                  Sign Up
                </button>
              </>
            )}
            <div className="header__dropdown-menu">
              <div
                styleName="header__dropdown-profile"
                onClick={handleProfileDropdown}
              >
                <Profile isLoggedIn={isLoggedIn} />
              </div>
              <div styleName="header__dropdown-dropdown">
                {isProfileDropdownOpen && isLoggedIn && (
                  <ProfileDropdown dropdown={"dropdown"} />
                )}
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  );
};

export default CSSModules(Navbar, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "log",
});
