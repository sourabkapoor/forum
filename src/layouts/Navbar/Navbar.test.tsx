import React from "react";
import {
  getByText,
  queryByPlaceholderText,
  queryByRole,
  queryByText,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import Navbar from "./Navbar";
import StoreProvider from "../../redux/provider";
import { MemoryRouter } from "react-router-dom";
import * as authSlice from "../../features/auth/authSlice";
import * as hooks from "../../hooks/hooks";

describe("Navbar", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <StoreProvider>
          <Navbar toggleDrawer={() => null} />
        </StoreProvider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("clicking Home tab opens commmunity tab", async () => {
    render(
      <MemoryRouter>
        <StoreProvider>
          <Navbar />
        </StoreProvider>
      </MemoryRouter>
    );

    const user = userEvent.setup();

    const communityDropdown = screen.getByText("Home");

    await user.click(communityDropdown);

    expect(screen.getByText("Create Community")).toBeInTheDocument();
  });
  test("clicking Log in opens Modal", async () => {
    render(
      <MemoryRouter>
        <StoreProvider>
          <Navbar />
        </StoreProvider>
      </MemoryRouter>
    );

    jest.spyOn(authSlice, "toggleSignInModal");
    const user = userEvent.setup();

    const logInButton = screen.queryByText("Log In");

    await user.click(logInButton!);

    expect(authSlice.toggleSignInModal).toHaveBeenCalledTimes(1);
  });
  // test("clicking Sign Up opens Modal", async () => {
  //   render(
  //     <MemoryRouter>
  //       <StoreProvider>
  //         <Navbar />
  //       </StoreProvider>
  //     </MemoryRouter>
  //   );

  //   jest.spyOn(authSlice, "toggleSignUpModal");
  //   const user = userEvent.setup();

  //   const logInButton = screen.queryByText("Sign Up");

  //   await user.click(logInButton!);

  //   expect(authSlice.toggleSignUpModal).toHaveBeenCalledTimes(1);
  // });
});
