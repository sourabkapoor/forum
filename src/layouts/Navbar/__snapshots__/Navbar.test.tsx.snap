// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="header__content"
    >
      <div
        class="header__left"
      >
        <a
          class="header__links"
          href="/"
        >
          <img
            alt="reddit logo"
            class="header__logo"
            src="logo.svg"
          />
          <img
            alt="the name reddit"
            class="header__logo-name"
            src="white-logo-name.svg"
          />
        </a>
        <div
          class="header__dropdown-container"
        >
          <div
            class="header__dropdown"
          >
            <div
              class="header__dropdown-left"
            >
              <svg
                class="header__dropdown-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
                />
              </svg>
              <h1
                class="header__dropdown-title"
              >
                Home
              </h1>
            </div>
            <div
              class="header__dropdown-right"
            >
              <svg
                class="header__dropdown-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                />
              </svg>
            </div>
          </div>
          <div
            class="header__dropdown-menu"
          />
        </div>
      </div>
      <div
        class="header__container"
      >
        <div
          class="header__search-icon-container"
        >
          <svg
            class="header__search-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M18.031 16.617l4.283 4.282-1.415 1.415-4.282-4.283A8.96 8.96 0 0 1 11 20c-4.968 0-9-4.032-9-9s4.032-9 9-9 9 4.032 9 9a8.96 8.96 0 0 1-1.969 5.617zm-2.006-.742A6.977 6.977 0 0 0 18 11c0-3.868-3.133-7-7-7-3.868 0-7 3.132-7 7 0 3.867 3.132 7 7 7a6.977 6.977 0 0 0 4.875-1.975l.15-.15z"
              />
            </g>
          </svg>
        </div>
        <input
          class="header__search-input"
          placeholder="Search Reddit"
        />
      </div>
      <div
        class="header__right"
      >
        <div
          class="header__buttons"
        >
          <button
            class="header__button header__button_type_log-in"
          >
            Log In
          </button>
          <button
            class="header__button header__button_type_sign-up"
          >
            Sign Up
          </button>
          <div
            class="header__dropdown-menu"
          >
            <div
              class="header__dropdown-profile"
            >
              <div
                class="profile"
              >
                <button
                  class="profile__button"
                >
                  <svg
                    class="profile__button-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M313.6 304c-28.7 0-42.5 16-89.6 16-47.1 0-60.8-16-89.6-16C60.2 304 0 364.2 0 438.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-25.6c0-74.2-60.2-134.4-134.4-134.4zM400 464H48v-25.6c0-47.6 38.8-86.4 86.4-86.4 14.6 0 38.3 16 89.6 16 51.7 0 74.9-16 89.6-16 47.6 0 86.4 38.8 86.4 86.4V464zM224 288c79.5 0 144-64.5 144-144S303.5 0 224 0 80 64.5 80 144s64.5 144 144 144zm0-240c52.9 0 96 43.1 96 96s-43.1 96-96 96-96-43.1-96-96 43.1-96 96-96z"
                    />
                  </svg>
                  <svg
                    class="profile__button-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="header__dropdown-dropdown"
            />
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
